### ISSUE TITLE
Support Input Union Types?

### ISSUE THREAD

**Issue by Jonatthu** [graphql-dotnet/graphql-dotnet#2055]
Not sure if this is part of graphql specs
But my use case is when I am creating an new user and I have a Foreign Key or create new child.

```
CreateUser {
  Username: String!
  AddressId: String
  Address: UserAddress
}
```

I would like to have it instead

```
CreateUser {
  Username: String!
  Address: String! | UserAddress!
}
```

Or also

```
CreateUser {
  Username: String!
  Address: UserAddressId! | UserAddress!
}

UserAddress {
  Street: String!
  State: String!
}

UserAddressId: {
  Id: String!
}

```

Is this possible already or something that could be implemented or actually is something can not be implemented because of the spec?

**Comment by sungam3r** [comment]
https://github.com/graphql/graphql-spec/blob/master/rfcs/InputUnion.md - work is under progress.

**Comment by Jonatthu** [comment]
@sungam3r 
Are we going for this one?
https://github.com/graphql/graphql-spec/blob/main/rfcs/InputUnion.md#-7-tagged-type

**Comment by sungam3r** [comment]
I observe from the part of the work on this problem. While still not part of the specification, so it is out of my interests yet.

**Comment by Shane32** [comment]
We should be able to write support for the `@oneOf` directive on top of the existing framework.  It's simply a custom directive coupled with a validation rule that asserts that only a single field has been specified.  With GraphQL 4.0 we can validate input fields and variables easily to ensure the requirement is met.  Plus with the experimental/draft directive introspection fields, introspection requests can examine input field types to identify which types are defined as having the `@oneOf` directive.

See: https://github.com/graphql/graphql-spec/pull/825

The new directive and validation rule can go with the custom `@length` directive we already have.

I think this would be great to include with 4.1; it should not be difficult if someone wants to do it.  I don't personally have a need for it though.

**Comment by sungam3r** [comment]
I understand but... things may change so I would wait for official words in the spec (at least approved/merged RFC).

**Comment by Shane32** [comment]
Having this directive exist is fully spec-compliant, just like how `@length` is.  And moreso than the directive introspection feature we added, which is not spec compliant.  So I'm just saying this is very unlike adding a new `UnionInput` type, which would be illegal according to the spec.  It's just an arbitrary feature.

Now if there is continued effort on the RFC, then sure, let's wait.  (I see updates within the last week or two 👍 .)  But often these RFC drafts just sit forever, seemingly forgotten.  And if that is the case, I would not wait to implement it, just as we did not wait for an RFC for `@length`.

If and when it becomes part of the spec, we will likely add a new class to the library (e.g. `InputUnionObjectGraphType`) and add it as a built-in feature.  If we already have added `@oneOf` to the library by then, we can mark the old directive as `[Obsolete]` -- or just update the behavior to match the spec.

**Comment by sungam3r** [comment]
I agree, I just consider this task with a reduced priority (since not even a RFC yet), there is a lot of other work.

**Comment by Socolin** [comment]
Hello,

I'm also interested by this feature

**Comment by Shane32** [comment]
@Socolin If you're referring to OneOf Input Objects, see https://github.com/graphql/graphql-spec/pull/825 .  Once the RFC is merged we will work on changes in this library for native support.

**Comment by sungam3r** [comment]
Current status of https://github.com/graphql/graphql-spec/pull/825 - not merged.
